src/python/pants/jvm/resolve/jvm_tool.py

```diff
@@ -161,5 +162,43 @@ async def gather_coordinates_for_jvm_lockfile(
     return ArtifactRequirements(requirements)
 
 
+@frozen_after_init
+@dataclass
+class ValidatedJvmToolLockfileRequest:
+
+    options_scope: str
+    artifact_inputs: FrozenOrderedSet[str]
+
+    def __init__(self, tool: JvmToolBase):
+        self.options_scope = tool.options_scope
+        self.artifact_inputs = FrozenOrderedSet(tool.artifact_inputs)
+        self.lockfile = tool.resolved_lockfile()
+
+    def __hash__(self):
+        return hash((self.options_scope, self.artifact_inputs, self.lockfile))
+
+
+@rule(desc="Validate JVM lockfile")
+async def validate_jvm_lockfile(
+    request: ValidatedJvmToolLockfileRequest,
+) -> CoursierResolvedLockfile:
+
+    lockfile = request.lockfile
+    requirements = await Get(
+        ArtifactRequirements,
+        GatherJvmCoordinatesRequest(
+            request.artifact_inputs, f"[{request.options_scope}].artifacts"
+        ),
+    )
+
+    if lockfile.metadata and not lockfile.metadata.is_valid_for(requirements):
+        raise ValueError(
+            f"Lockfile for {request.options_scope} was generated with different "
```

Comment:
Error handling code needs to be better (at least as good as Python's :))

Reply:
For context, the Python tool lockfile error message code is a [hundred lines of code](https://github.com/pantsbuild/pants/blob/main/src/python/pants/backend/python/util_rules/pex.py#L563-L677) with carefully specified error messages. We should probably tackle this as a separate issue, as this PR is already pretty big.